### PR TITLE
WIP: Exercise 3 

### PR DESCRIPTION
**Pull request by aloparev**
PR review can be found [here](https://github.com/Systems-Development-and-Frameworks/bgrakia/pull/10).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+.eslintrc.back
+backend/package-lock.json
+backend/yarn.lock
+backend/node_modules
```
Comment (MaykAkifovski) [inline .gitignore:4]: If you like, you can also create local ,gitignore in backend

File: backend/src/DataSource/posts-data-source.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        return post
+      }
+      else{
+        throw new UserInputError("This user voted on this post already", {invalidArgs: [currUser]});
+      }
+    }
+    else{
+      throw new UserInputError("No post with this ID", {invalidArgs: [id]});
+    }
+  }
```
Comment (MaykAkifovski) [inline backend/src/DataSource/posts-data-source.js:67]: For me it's a duplication of code here. Maybe you can implement one function and call it with parameter +1 for upvote and -1 for downvote. What do you think?

File: backend/src/DataSource/users-data-source.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+class UsersDataSource extends RESTDataSource {
+  
+  constructor() {
+    super();
+    this.users = []
+  }
+
+  initialize({context}) {
+    //console.log('UsersDataSource: ', context)
```
Comment (MaykAkifovski) [inline backend/src/DataSource/users-data-source.js:17]: it's a bad practice to commit commented code

**Review by MaykAkifovski — APPROVED** [review]
Overall LGTM. I have commented some bad practices of Clean Code perspective. I am not sure what happened with your code, but all the components in your Frontend are listed as new created. Maybe its because of the new packaging-naming. If it was written in lower-case earlier, you should stick to kebab-case.

**Comment by ilonae** [comment]
@roschaefer @medizinmensch Right now, our checks for the front end fail because we've decided to move on and implement optional features such as Apollo Client. Hope you can acknowledge that, it's something we'd like to fix in the future.

**Review by medizinmensch — APPROVED** [review]
Hey Group! Good job!

Good Job connected your webapp with a client! Very cool to see!

The only thing missing is the review for the other group. :x:
Do you have a link for that?

Otherwise a nice 9/10 ⭐

**Comment by ilonae** [comment]
> Hey Group! Good job!
> 
> Good Job
> 
> See my comments below.
> 
> @ilonae , it's fine that the webapp tests fail for the moment. Amazing that you already (started? to) connected your webapp with a client! Very cool to see!
> 
> The only thing missing is the review for the other group. x
> Do you have a link for that?
> 
> Otherwise a nice 9/10 star

Yes, sorry about that. @aloparev made a review for the other group, it can be found here: https://github.com/Systems-Development-and-Frameworks/bgrakia/pull/10 .

**Comment by roschaefer** [comment]
> Hey Group! Good job!
> 
> Good Job
> 
> See my comments below.
> 
> @ilonae , it's fine that the webapp tests fail for the moment. Amazing that you already (started? to) connected your webapp with a client! Very cool to see!
> 
> The only thing missing is the review for the other group. 
> Do you have a link for that?
> 
> Otherwise a nice 9/10

I gave the remaining :star: for this, well done :+1:

**Pull request merged**